crates/astria-sequencer/src/proposal/commitment.rs

```diff
@@ -0,0 +1,81 @@
+use std::collections::BTreeMap;
+
+use anyhow::{
+    Context as _,
+    Error,
+    Result,
+};
+use bytes::Bytes;
+use tendermint::merkle::simple_hash_from_byte_vectors;
+
+use crate::transaction::Signed;
+
+/// Called when we receive a `PrepareProposal` or `ProcessProposal` consensus message.
+///
+/// In the case of `PrepareProposal`, we use this function to generate the `commitment_tx`
+/// to be placed at the start of the block.
+///
+/// In the case of `ProcessProposal`, we use this function to generate and verify the
+/// `commitment_tx` expected at the start of the block.
+///
+/// This function sorts the block's `secondary::Action`s contained within the transactions
+/// using their `chain_id`. It then returns the merkle root of the tree where each leaf is
+/// a commitment of `secondary::Action`s with the same `chain_id`. The leaves are ordered
+/// by namespace in ascending order, where `namespace(chain_id) = Sha256(chain_id)[0..10]`.
+/// This structure can be referred to as the "action tree".
+///
+/// The leaf, which contains a commitment to every action with the same `chain_id`, is currently
+/// implemented as ( `namespace(chain_id)` || root of merkle tree of the `secondary::Action`s ).
+/// This is somewhat arbitrary, but could be useful for proof of an action within the action tree.
+pub(crate) fn generate_transaction_commitment(txs_bytes: &[Bytes]) -> Result<[u8; 32]> {
+    let txs = txs_bytes
+        .iter()
+        .map(|tx_bytes| Signed::try_from_slice(tx_bytes))
+        .collect::<Result<Vec<Signed>, Error>>()
+        .context("failed to deserialize transactions")?;
+
+    let chain_id_to_txs = sort_txs_by_chain_id(&txs);
+
+    let mut leaves: Vec<Vec<u8>> = vec![];
```

Comment:
Could we get a comment above this about what the leaves are structurally (ie not u8, those bytes represent something)